test/4_test.py

```diff
@@ -0,0 +1,315 @@
+                headers = {
+                    'Authorization': self.access_token
+                }
+            ).json()
+            good_events[response['event_id']] = "Good image with good description, type %s" % message_type
+
+            print('Send good image with bad description, it should be rejected')
+            response = requests.put('http://localhost:8080/_matrix/client/r0/rooms/%s/send/m.room.message/gibd-%s' % (quoted_room_id, uuid.uuid1()),
+                json = {
+                    'body': 'A text with a hidden link to evil.example.com',
+                    'msgtype': message_type,
+                    'url': good_mxid,
+                    'info': {
+                        'w': 320,
+                        'h': 200,
+                        'size': len(good_file_content),
+                    }
+                },
+                headers = {
+                    'Authorization': self.access_token
+                }
+            ).json()
+            if response.get('event_id', None) is None:
+                print('Message was rejected immediately')
+            else:
+                # Message may be redacted later
+                bad_events[response['event_id']] = "Good image with bad description, type %s" % message_type
+
+            print('Send bad image with good description, it should be rejected')
+            response = requests.put('http://localhost:8080/_matrix/client/r0/rooms/%s/send/m.room.message/bigd-%s' % (quoted_room_id, uuid.uuid1()),
+                json = {
+                    'body': 'A text without any link',
+                    'msgtype': message_type,
+                    'url': evil_mxid,
+                    'info': {
+                        'w': 320,
+                        'h': 200,
+                        'size': len(evil_file_content),
+                    }
+                },
+                headers = {
+                    'Authorization': self.access_token
+                }
+            ).json()
+            if response.get('event_id', None) is None:
+                print('Message was rejected immediately')
+            else:
+                # Message may be redacted later
+                bad_events[response['event_id']] = "Good image with bad description, type %s" % message_type
+
+        print('Sending canary event, to ensure that all previous events have been flushed')
+        response = requests.put('http://localhost:8080/_matrix/client/r0/rooms/%s/send/m.room.message/canary-%s' % (quoted_room_id, uuid.uuid1()),
+            json = {
+                'body': 'A text without any link',
+                'format': 'org.matrix.custom.html',
+                'formatted_body': 'A rich text without any link',
+                'msgtype': 'm.text'
+            },
+            headers = {
+                'Authorization': self.access_token
+            }
+        ).json()
+        canary_event = response['event_id']
+
+        # Wait until the canary has been treated.
```

Comment:
I'm a bit confused at why we need to do this loop twice, why can't it be done once and just continue until the canary is found?

Reply:
That's in case the server doesn't immediately redact bad events.
Adding some documentation.